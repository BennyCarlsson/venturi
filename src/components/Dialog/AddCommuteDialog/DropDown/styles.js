import styled from "styled-components";
import Colors from "tokens/colors";

export const Ul = styled.ul`
  position: absolute;
  z-index: 2;
  width: 100%;
  background-color: ${Colors.white};
  list-style: none;
  margin: 0;
  padding: 8px;
  border: 1px solid #726a95;
`;

export const Li = styled.li`
  padding: 8px 0;
  cursor: pointer;
  &:hover {
    background-color: #fafafa;
  }
`;
