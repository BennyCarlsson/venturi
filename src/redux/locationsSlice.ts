import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'
import { getLocationOName } from 'api/vasttrafik/vasttrafik'
import { LocationList } from 'types'

const _getLocationOName = async (
  name: string,
  { signal }: { signal: AbortSignal }
) => {
  signal.addEventListener('abort', () => {
    console.log('cancelled')
  })
  return getLocationOName(name, signal)
}

export const fetchDestinationLocationsOnName = createAsyncThunk(
  'locations/fetchDestinationLocationsOnName',
  _getLocationOName
)

export const fetchOriginLocationsOnName = createAsyncThunk(
  'locations/fetchOriginLocationsOnName',
  _getLocationOName
)

export enum LocationType {
  StopLocation,
  CoordLocation,
}

export type Location = {
  id: string
  idx: string
  name: string
  lat: string
  lon: string
  type: LocationType
}

// Todo extract and write tests
const convertData = (data: LocationList): Location[] | undefined => {
  console.log('data', data)
  let stopLocations: Location[] = []
  if (data.StopLocation) {
    if (!Array.isArray(data.StopLocation)) {
      data.StopLocation = [data.StopLocation]
    }
    stopLocations = data.StopLocation?.map((loc) => ({
      id: loc.id,
      idx: loc.idx,
      name: loc.name,
      lat: loc.lat,
      lon: loc.lon,
      type: LocationType.StopLocation,
    }))
  }

  let coordLocation: Location[] = []
  if (data.CoordLocation) {
    if (!Array.isArray(data.CoordLocation)) {
      data.CoordLocation = [data.CoordLocation]
    }
    coordLocation = data.CoordLocation?.map((loc) => ({
      id: loc.name,
      idx: loc.idx,
      name: loc.name,
      lat: loc.lat,
      lon: loc.lon,
      type: LocationType.CoordLocation,
    }))
  }
  return stopLocations
    .concat(coordLocation)
    .sort((a, b) => parseInt(a.idx) - parseInt(b.idx))
}

export interface TripState {
  loadingOrigin: boolean
  originLocations?: Location[]
  originError?: string
  loadingDestination: boolean
  destinationLocations?: Location[]
  destinationError?: string
}

const initialState: TripState = {
  loadingOrigin: false,
  loadingDestination: false,
}

const locationSearchSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchOriginLocationsOnName.pending, (state, action) => {
      state.loadingOrigin = true
    })
    builder.addCase(fetchOriginLocationsOnName.fulfilled, (state, action) => {
      state.loadingOrigin = false
      if (action.payload.LocationList.error) {
        state.originError = action.payload.LocationList.errorText
      } else {
        state.originLocations = convertData(action.payload.LocationList)
      }
    })
    builder.addCase(
      fetchDestinationLocationsOnName.pending,
      (state, action) => {
        state.loadingDestination = true
      }
    )
    builder.addCase(
      fetchDestinationLocationsOnName.fulfilled,
      (state, action) => {
        state.loadingDestination = false
        if (action.payload.LocationList.error) {
          state.destinationError = action.payload.LocationList.errorText
        } else {
          state.destinationLocations = convertData(action.payload.LocationList)
        }
      }
    )
  },
})

export default locationSearchSlice.reducer
