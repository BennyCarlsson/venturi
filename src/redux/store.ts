import { configureStore } from '@reduxjs/toolkit'
import locationsReducer from './locationsSlice'
import tripReducer from './tripSlice'
import tripListReducer from './tripListSlice'
import dialogReducer from './dialogSlice'
import { loadOriginAndDestination, loadTriplList } from '../utils/localStorage'

const localStorageDataTrip = loadOriginAndDestination()
const localStorageDataTripList = loadTriplList()

const store = configureStore({
  reducer: {
    trip: tripReducer,
    tripList: tripListReducer,
    locations: locationsReducer,
    dialog: dialogReducer,
  },
  preloadedState:
    localStorageDataTrip && localStorageDataTripList
      ? {
          trip: {
            origin: localStorageDataTrip.origin,
            destination: localStorageDataTrip.destination,
          },
          tripList: localStorageDataTripList,
        }
      : {},
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export default store
