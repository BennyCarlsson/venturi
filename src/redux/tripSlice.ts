import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { getTrip } from 'api/vasttrafik/vasttrafik';

type tripOrderType = {
  fromId: string;
  toId: string;
};
export const fetchTrip = createAsyncThunk('trip/fetchTrip', async (tripOrder: tripOrderType) => {
  const response = await getTrip(tripOrder.fromId, tripOrder.toId);
  return response;
});

export const tripSlice = createSlice({
  name: 'test',
  initialState: {
    value: 0
  },
  reducers: {
    increment: (state) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      console.log('hello');

      state.value += 1;
    },
    decrement: (state) => {
      state.value -= 1;
    },
    incrementByAmount: (state, action) => {
      state.value += action.payload;
    }
  }
});

// Action creators are generated for each case reducer function
export const { increment, decrement, incrementByAmount } = tripSlice.actions;

export default tripSlice.reducer;
