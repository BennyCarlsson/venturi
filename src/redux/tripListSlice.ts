import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Location } from './locationsSlice'

export interface TripListState {
  trips: { name: string; origin: Location; destination: Location }[]
}

const initialState: TripListState = { trips: [] }

const tripListSlice = createSlice({
  name: 'trip',
  initialState,
  reducers: {
    addTripToList: (
      state,
      action: PayloadAction<{
        name?: string
        origin: Location
        destination: Location
      }>
    ) => {
      state.trips.unshift({
        name: action.payload.name || action.payload.origin.name,
        origin: action.payload.origin,
        destination: action.payload.destination,
      })
    },
    moveTripToFirst: (state, action: PayloadAction<{ index: number }>) => {
      const index = action.payload.index
      const trip = state.trips[index]
      state.trips.splice(index, 1)
      state.trips.unshift(trip)
      return state
    },
  },
})

export const { addTripToList, moveTripToFirst } = tripListSlice.actions

export default tripListSlice.reducer
