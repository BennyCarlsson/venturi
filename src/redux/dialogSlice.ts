import { createSlice } from '@reduxjs/toolkit'
import { RootState } from './store'

export type DialogType = {
  show: boolean
  isIntro: boolean
}
const initialState: DialogType = {
  show: false,
  isIntro: false,
}
const dialogSlice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    showDialog: (state) => {
      state.show = true
      state.isIntro = false
    },
    showIntroDialog: (state) => {
      state.show = true
      state.isIntro = true
    },
    hideDialog: (state) => {
      state.show = false
      state.isIntro = false
    },
  },
})

export const { showDialog, showIntroDialog, hideDialog } = dialogSlice.actions

export const selectShowDialog = (state: RootState) => state.dialog.show
export const selectIsIntro = (state: RootState) => state.dialog.isIntro
export default dialogSlice.reducer
