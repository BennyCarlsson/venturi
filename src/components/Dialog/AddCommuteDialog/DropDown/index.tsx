import { Fragment } from 'react';
import { Location } from 'redux/locationsSlice';
import { Li, Ul } from './styles';

type DropDownProps = {
  locations: Location[] | undefined;
  setInput: (value: React.SetStateAction<string>) => void;
  setLocation: (value: React.SetStateAction<Location | null>) => void;
  show: boolean;
};

const DropDown = ({ locations, setInput, setLocation, show }: DropDownProps) => {
  const shouldRender = () => show && Array.isArray(locations) && locations.length > 0;
  return (
    <Fragment>
      {shouldRender() && (
        <Ul>
          {locations?.map((loc) => (
            <Li
              key={`key-${loc.id}`}
              onClick={(e) => {
                e.preventDefault();
                setInput(loc.name);
                setLocation(loc);
              }}
            >
              {loc.name}
            </Li>
          ))}
        </Ul>
      )}
    </Fragment>
  );
};

export default DropDown;
