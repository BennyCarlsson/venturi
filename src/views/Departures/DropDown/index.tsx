import { Fragment } from 'react';
import { Location } from 'redux/locationsSlice';
import { Li, Ul } from './styles';

type DropDownProps = {
  locations: Location[] | undefined;
  setInput: (value: React.SetStateAction<string>) => void;
  setLocation: (value: React.SetStateAction<Location | null>) => void;
  show: boolean;
};

const DropDown = ({ locations, setInput, setLocation, show }: DropDownProps) => {
  return (
    <Fragment>
      {show && (
        <Ul>
          {locations?.map((loc) => (
            <Li
              key={`key-${loc.id}`}
              onClick={(e) => {
                e.preventDefault();
                console.log(loc.id);
                setInput(loc.name);
                setLocation(loc);
              }}
            >
              {loc.name}
            </Li>
          ))}
        </Ul>
      )}
    </Fragment>
  );
};

export default DropDown;
