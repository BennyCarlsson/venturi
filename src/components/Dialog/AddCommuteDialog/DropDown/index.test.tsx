import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { LocationType } from 'redux/locationsSlice';
import DropDown from '.';
const locations = [
  {
    id: '1',
    idx: '1',
    name: 'name',
    lat: '1',
    lon: '1',
    type: LocationType.CoordLocation
  },
  {
    id: '2',
    idx: '2',
    name: 'name2',
    lat: '2',
    lon: '2',
    type: LocationType.CoordLocation
  },
  {
    id: '3',
    idx: '3',
    name: 'name3',
    lat: '3',
    lon: '3',
    type: LocationType.CoordLocation
  }
];

describe('<DropDown />', () => {
  const spySetInput = jest.fn();
  const spySetLocation = jest.fn();
  beforeEach(() => {
    render(
      <DropDown locations={locations} setInput={spySetInput} setLocation={spySetLocation} show />
    );
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(<DropDown locations={locations} setInput={jest.fn()} setLocation={jest.fn()} show />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    expect(screen).toBeTruthy();
  });

  it('should render 3 li', () => {
    expect(screen.queryByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem')).toHaveLength(3);
  });

  it('should click the first listitem', () => {
    userEvent.click(screen.getAllByRole('listitem')[0]);
    expect(spySetInput).toHaveBeenCalledTimes(1);
    expect(spySetInput).toHaveBeenCalledWith(locations[0].name);
    expect(spySetLocation).toHaveBeenCalledTimes(1);
    expect(spySetLocation).toHaveBeenCalledWith(locations[0]);
  });

  it('should click the second listitem', () => {
    userEvent.click(screen.getAllByRole('listitem')[1]);
    expect(spySetInput).toHaveBeenCalledTimes(1);
    expect(spySetInput).toHaveBeenCalledWith(locations[1].name);
    expect(spySetLocation).toHaveBeenCalledTimes(1);
    expect(spySetLocation).toHaveBeenCalledWith(locations[1]);
  });

  it('should not render if show is false', () => {
    cleanup();
    render(
      <DropDown
        locations={locations}
        setInput={spySetInput}
        setLocation={spySetLocation}
        show={false}
      />
    );
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });

  // location undefined
  it('should not render if location is false', () => {
    cleanup();
    render(
      <DropDown locations={undefined} setInput={spySetInput} setLocation={spySetLocation} show />
    );
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });

  it('should not render if location is empty array', () => {
    cleanup();
    render(<DropDown locations={[]} setInput={spySetInput} setLocation={spySetLocation} show />);
    expect(screen.queryByRole('list')).not.toBeInTheDocument();
  });
});
