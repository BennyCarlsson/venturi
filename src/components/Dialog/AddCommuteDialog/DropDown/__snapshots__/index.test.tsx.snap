// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropDown /> should match snapshot 1`] = `
<ul
  className="sc-bdfBwQ hqbaFS"
>
  <li
    className="sc-gsTCUz iMZhDb"
    onClick={[Function]}
  >
    name
  </li>
</ul>
`;
